File: programmers/달리기 경주.py
# 달리기 경주 - level 1
# https://school.programmers.co.kr/learn/courses/30/lessons/178871

# 내 답안 O(N^2)
def solution(players, callings):
    for c in callings:
        for p in range(len(players)):
            if players[p] == c:
                players[p-1], players[p] = players[p], players[p-1]
    
    answer = players
    return answer

# 다른 사람 답안 O(N)
# 딕셔너리 공부해야겠다
def solution2(players, callings):
    p_dic = {player: index for index, player in enumerate(players)} # index = 해당 선수의 현재 순위

    for p in callings:
        c = p_dic[p]

        p_dic[p] -= 1 # calling된 선수 순위 - 1
        p_dic[players[c-1]] += 1 # calling된 선수 바로 앞 선수 순위 + 1
        players[c-1], players[c] = players[c], players[c-1] # 두 선수 위치 변경
    
    return players

print(solution(["mumu", "soe", "poe", "kai", "mine"], ["kai", "kai", "mine", "mine"]))

File: programmers/Level1/달리기 경주.js
// Level 1. 달리기 경주 게임
// https://school.programmers.co.kr/learn/courses/30/lessons/178871?language=javascript

function solution(players, callings) {
  // 달리기 경주
  // 바로 앞의 선수 추월할 때 추월한 선수 이름 부른다
  // 1등부터 현재 등수 순서대로 담긴 문자열 배열 players
  // 경주 끝났을 때 선수들의 이름을 1등부터 등수 순서대로 배열 담아 return

  let ranks = {};
  players.forEach((player, i) => (ranks[player] = i));

  callings.forEach((player) => {
    let front_player = players[ranks[player] - 1];
    ranks[player] -= 1;
    ranks[front_player] += 1;

    players[ranks[player]] = player;
    players[ranks[front_player]] = front_player;
    // O(N)의 시간 걸리는 아래의 for문 대신 위의 hash 사용
    // for (const [key, value] of Object.entries(ranks)) {
    //     if(key !== player && value === ranks[player])
    //         ranks[key] += 1;
    // }
  });

  let answer = [];
  players.forEach((r) => answer.push(''));
  for (const [key, value] of Object.entries(ranks)) {
    answer[value] = key;
  }

  return answer;
}
